src/service/cid-impl.js

```diff
@@ -121,32 +121,56 @@ export class Cid {
     return consent.then(() => {
       return viewerForDoc(this.ampdoc).whenFirstVisible();
     }).then(() => {
-      return getExternalCid(this, getCidStruct,
-          opt_persistenceConsent || consent);
+      const cidPromise = this.getExternalCid_(
+          getCidStruct, opt_persistenceConsent || consent);
+      // Getting the CID might involve an HTTP request. We timeout after 10s.
+      return timerFor(this.ampdoc.win)
+          .timeoutPromise(10000, cidPromise,
+              `Getting cid for "${getCidStruct.scope}" timed out`)
+          .catch(error => {
+            rethrowAsync(error);
+          });
     });
   }
-}
 
-/**
- * Returns the "external cid". This is a cid for a specific purpose
- * (Say Analytics provider X). It is unique per user, that purpose
- * and the AMP origin site.
- * @param {!Cid} cid
- * @param {!GetCidDef} getCidStruct
- * @param {!Promise} persistenceConsent
- * @return {!Promise<?string>}
- */
-function getExternalCid(cid, getCidStruct, persistenceConsent) {
-  /** @const {!Location} */
-  const url = parseUrl(cid.ampdoc.win.location.href);
-  if (!isProxyOrigin(url)) {
-    return getOrCreateCookie(cid, getCidStruct, persistenceConsent);
+  /**
+   * Returns the "external cid". This is a cid for a specific purpose
+   * (Say Analytics provider X). It is unique per user, that purpose
+   * and the AMP origin site.
+   * @param {!GetCidDef} getCidStruct
+   * @param {!Promise} persistenceConsent
+   * @return {!Promise<?string>}
+   */
+  getExternalCid_(getCidStruct, persistenceConsent) {
+    /** @const {!Location} */
+    const url = parseUrl(this.ampdoc.win.location.href);
+    if (!isProxyOrigin(url)) {
+      return getOrCreateCookie(this, getCidStruct, persistenceConsent);
+    }
+    const viewer = viewerForDoc(this.ampdoc);
+    if (viewer.hasCapability('cid')) {
+      return this.getScopedCidFromViewer_(getCidStruct.scope);
+    }
+    return getBaseCid(this, persistenceConsent)
+        .then(baseCid => {
+          return cryptoFor(this.ampdoc.win).sha384Base64(
+              baseCid + getProxySourceOrigin(url) + getCidStruct.scope);
+        });
+  }
+
+  /**
+   * @param {!string} scope
+   * @return {!Promise<?string>}
+   */
+  getScopedCidFromViewer_(scope) {
+    const viewer = viewerForDoc(this.ampdoc);
+    return viewer.isTrustedViewer().then(trusted => {
+      if (!trusted) {
+        return;
```

Comment:
> localStorage is better than nothing (works in Chrome, Firefox, works for session in Safari), so I think falling back makes sense.

wanted to keep the logic the same as the old API, which does not fallback

> But only trusted viewers should tell us they have the cid capability, so I think we should throw here! (Async?) so that we detect if this goes systematically wrong.

Yes, we can throw here. This is an unexpected stage.